File: ugly-number-ii.cc
// Ugly Number II
class Solution {
public:
  int nthUglyNumber(int n) {
    vector<int> a{1};
    int i = 0, j = 0, k = 0;
    while (a.size() < n) {
      int x = min(min(a[i]*2, a[j]*3), a[k]*5);
      a.push_back(x);
      if (a[i]*2 == x) i++;
      if (a[j]*3 == x) j++;
      if (a[k]*5 == x) k++;
    }
    return a[n-1];
  }
};

File: ugly-number-ii.rs
impl Solution {
    pub fn nth_ugly_number(n: i32) -> i32 {
        let mut a = vec![1];
        let mut i2 = 0;
        let mut i3 = 0;
        let mut i5 = 0;
        while a.len() < n as usize {
            let mn = (a[i2]*2).min(a[i3]*3).min(a[i5]*5);
            a.push(mn);
            if a[i2]*2 == mn { i2 += 1; }
            if a[i3]*3 == mn { i3 += 1; }
            if a[i5]*5 == mn { i5 += 1; }
        }
        *a.last().unwrap()
    }
}
